subprojects/provider/src/main/kotlin/org/gradle/kotlin/dsl/codegen/ApiTypeProvider.kt

```diff
@@ -0,0 +1,528 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.kotlin.dsl.codegen
+
+import org.gradle.api.Incubating
+
+import org.gradle.kotlin.dsl.accessors.primitiveTypeStrings
+import org.gradle.kotlin.dsl.support.ClassBytesRepository
+import org.gradle.kotlin.dsl.support.classPathBytesRepositoryFor
+
+import org.jetbrains.org.objectweb.asm.AnnotationVisitor
+import org.jetbrains.org.objectweb.asm.Attribute
+import org.jetbrains.org.objectweb.asm.ClassReader
+import org.jetbrains.org.objectweb.asm.ClassReader.SKIP_CODE
+import org.jetbrains.org.objectweb.asm.ClassReader.SKIP_DEBUG
+import org.jetbrains.org.objectweb.asm.ClassReader.SKIP_FRAMES
+import org.jetbrains.org.objectweb.asm.FieldVisitor
+import org.jetbrains.org.objectweb.asm.Opcodes.ACC_PUBLIC
+import org.jetbrains.org.objectweb.asm.Opcodes.ACC_STATIC
+import org.jetbrains.org.objectweb.asm.Opcodes.ACC_SYNTHETIC
+import org.jetbrains.org.objectweb.asm.Opcodes.ASM6
+import org.jetbrains.org.objectweb.asm.Type
+import org.jetbrains.org.objectweb.asm.TypePath
+import org.jetbrains.org.objectweb.asm.signature.SignatureReader
+import org.jetbrains.org.objectweb.asm.signature.SignatureVisitor
+import org.jetbrains.org.objectweb.asm.tree.AnnotationNode
+import org.jetbrains.org.objectweb.asm.tree.ClassNode
+import org.jetbrains.org.objectweb.asm.tree.MethodNode
+
+import java.io.Closeable
+import java.io.File
+import java.util.ArrayDeque
+
+import javax.annotation.Nullable
+
+import kotlin.LazyThreadSafetyMode.NONE
+
+
+internal
+fun apiTypeProviderFor(
+    classPath: List<File>,
+    classPathDependencies: List<File> = emptyList(),
+    parameterNamesSupplier: ParameterNamesSupplier = { null }
+): ApiTypeProvider =
+
+    ApiTypeProvider(classPathBytesRepositoryFor(classPath, classPathDependencies), parameterNamesSupplier)
+
+
+private
+typealias ApiTypeSupplier = () -> ApiType
+
+
+internal
+typealias ParameterNamesSupplier = (String) -> List<String>?
+
+
+/**
+ * Provides [ApiType] instances by Kotlin source name from a class path.
+ *
+ * Keeps JAR files open for fast lookup, must be closed.
+ * Once closed, type graph navigation from [ApiType] and [ApiFunction] instances
+ * will throw [IllegalStateException].
+ *
+ * Limitations:
+ * - supports Java byte code only, not Kotlin
+ * - does not support nested Java arrays as method parameters
+ * - does not distinguish co-variance and contra-variance
+ */
+internal
+class ApiTypeProvider(
+    private val repository: ClassBytesRepository,
+    parameterNamesSupplier: ParameterNamesSupplier
+) : Closeable {
+
+    private
+    val context = Context(this, parameterNamesSupplier)
```

Comment:
Doesn't the compiler warn you about leaking a reference to `this` in the constructor here?
Couldn't this be solved with a `lazy` here?

Reply:
Nope, `Context` is an inner static class, see below